master/internal/stream/system.go

```diff
@@ -364,6 +344,42 @@ func prepareWebsocketMessage(obj stream.PreparableMessage) interface{} {
 	return msg
 }
 
+// startupMsgHandler reads and responds to startup messages sent by streaming clients.
+func startupMsgHandler(
+	ctx context.Context,
+	startupMsg StartupMsg,
+	ps *PublisherSet,
+	ss *SubscriptionSet,
+	user model.User,
+	streamer *stream.Streamer,
+) (msgs []interface{}, err error) {
+	defer func() {
+		syncMsg := SyncMsg{SyncID: startupMsg.SyncID}
+		msgs = append(msgs, syncMsg.toPreparedMessage())
+	}()
+
+	// check for active subscriptions
+	if ss != nil {
+		ss.UnsubscribeAll()
+	}
+
+	// create new subscription set
+	tempSS, err := NewSubscriptionSet(ctx, streamer, ps, user, startupMsg.Subscribe)
+	*ss = tempSS
+	if err != nil {
+		return msgs, errors.Wrap(err, "creating subscription set")
+	}
```

Comment:
I'm still not crazy about this, because you're deferring a method of an object in the parent, then replacing that object halfway through the deferred cleanup.  That's sketch.  I see you already protected against failure cases because ss.UnsubscribeAll() has nil pointer checks.  But that's not enough, you'd have to make sure that none of those pointers are ever non-nil but still unsafe to call .Unregister on.  If I read NewSubscriptionSet(), I see that is in fact the case.

That's a lot of reading to convince myself this isn't a bug.

I don't see a solution which is indisputably better, so instead let me suggest an alternative and you can pick which way you want to take it.  If your strategy is "surprising" because it replace a function before the `defer` of that function, my strategy is "surprising" because it unrolls the underlying for loop in a nonstandard way.


```go
// processStream processes a startup message, then streams live updates until either:
//   - another startup message arrives, in which case it returns it, or
//   - the streamer is closed gracefully, in which case it returns nil, or
//   - an error occurs
func (ps *PublisherSet) processStream(...) ([]StartupMsg, error) {
	// create new subscription set
	ss, err := NewSubscriptionSet(ctx, streamer, ps, user, startupMsg.Subscribe)
	if err != nil {
		return msgs, errors.Wrap(err, "creating subscription set")
	}
	defer ss.UnsubscribeAll()

	// startup subscription set
	offlineMsgs, err := ss.Startup(ctx, user, startupMsg)
	if err != nil {
		return msgs, errors.Wrapf(err, "gathering startup messages")
	}
	msgs = append(msgs, offlineMsgs...)

	// always include a sync message
	syncMsg := SyncMsg{SyncID: startupMsg.SyncID}
	msgs = append(msgs, syncMsg.toPreparedMessage())

	// write offline msgs to the websocket
	err = writeAll(socket, msgs)
	if err != nil {
		// XXX: don't log broken pipe errors.
		return errors.Wrapf(err, "error writing to socket")
	}

	// startup done, begin streaming of supported online events:
	//   - insertions
	//   - updates
	//   - deletions
	//   - fallin
	//   - fallout
	//
	// (note that online appearances and disappearances are not supported; we'll detect those
	// situations and just break the connection to streaming clients).

	// waitForSomething returns a tuple of (first-startup-msg, msgs, closed)
	waitForSomething := func() (*StartupMsg, []interface{}, bool) {
		streamer.Cond.L.Lock()
		defer streamer.Cond.L.Unlock()
		// NOTE: this is different!
		for len(*startups) == 0 && len(streamer.Msgs) == 0 && !streamer.Closed {
			streamer.Cond.Wait()
		}
		// steal outputs
		var startup *StartupMsg
		if len(*startups) == 0 {
			startup = (*startups)[0]
			*startups = (*startups)[1:]
		}
		msgs := streamer.Msgs
		streamer.Msgs = nil
		return startup, msgs, streamer.Closed
	}

	for {
		startup, msgs, closed := waitForSomething()

		// is the streamer closed?
		if closed {
			return nil, nil
		}

		// soft reset?
		if startup {
			return startup, nil
		}

		// otherwise write all the messages we just got
		err = writeAll(socket, msgs)
		if err != nil {
			// XXX: don't log broken pipe errors.
			return errors.Wrapf(err, "error writing to socket")
		}
	}
}

func (ps *PublisherSet) entrypoint(...) error {
	// get permission change channel
	var bootemChan chan struct{}
	func() {
		ps.bootLock.Lock()
		defer ps.bootLock.Unlock()
		bootemChan = ps.bootemChan
	}()

	streamer := stream.NewStreamer(prepareFunc)

	// read the first startup message
	var startupMsg StartupMsg
	err := socket.ReadJSON(&startupMsg)
	if err != nil {
		return errors.Wrapf(err, "error while reading initial startup message")
	}

	// startups is where we collect StartupMsg messages we read from the websocket until
	// waitForSomething() delivers those messages to the websocket goroutine.
	var startups []StartupMsg

	// always be reading for new startup messages
	go func() {
		defer streamer.Close()
		for {
			var mods StartupMsg
			err := socket.ReadJSON(&mods)
			if err != nil {
				if websocket.IsUnexpectedCloseError(
					err, websocket.CloseGoingAway, websocket.CloseAbnormalClosure,
				) {
					log.Errorf("unexpected close error: %v", err)
				}
				break
			}
			// wake up streamer goroutine with the newly-read StartupMsg
			func() {
				streamer.Cond.L.Lock()
				defer streamer.Cond.L.Unlock()
				streamer.Cond.Signal()
				// start
				streamer.startups = append(startups, mods)
			}()
		}
	}()

	// detect context cancelation, and bring it into the websocket thread
	go func() {
		select {
		case <-ctx.Done():
			streamer.Close()
		case <-ssupCtx.Done():
			// close streamer if supervisor is down
			streamer.Close()
		case <-bootemChan:
			// close this streamer if online appearance/disappearance occurred
			streamer.Close()
		}
	}()

	for {
		nextStartupMsg, err := ps.processStream(startupMsg, &startups, ...)
		if err != nil {
			// stream failed
			return err
		}
		if !nextStartupMsg {
			// stream closed
			return nil
		}
		// stream soft reset: the last subscription ended due to a user sending another StartupMsg
		startupMsg = *nextStartupMsg
	}
}
```

My solution differs in a couple of ways:
- it doesn't worry about the final online messages that arrive after a new StartupMsg; it's racy anyway, the client isn't allowed to care about how we resolve races.
- the defer applies to the lifetime that it looks like it applies to... no replacing things mid-liftime
- it is logically like using a while-true loop with a conditional break in either the top or bottom of the loop; it's kind of a weird way to unroll a loop.

Reply:
I like the idea of making sure that the life time of an object is easy to understand, and this explicitly ties in that concept, rather than having the same subscription set get mutated over the course of the connection, your approach makes it very clear which subscription set is being used (the same one during the entire `processStream`).

I like your solution better than mine if not for anything else, it requires much less reading to confirm correctness & maintain.